test/transforms.js

```diff
@@ -0,0 +1,142 @@
+      let fullBatchCounter = 0;
+      let partialBatchCounter = 0;
+      let elementCounter = 0;
+      let noOversizeBatches = true;
+      return pipeline(Readable.from(Array(elements).keys()), new BatchingStream(batchSize), new Writable({
+        objectMode: true,
+        write(chunk, encoding, callback) {
+          batchCounter++;
+          const currentBatchSize = chunk.length;
+          if (currentBatchSize < batchSize) {
+            // partial batch
+            partialBatchCounter++;
+          } else if (currentBatchSize === batchSize) {
+            // full batch
+            fullBatchCounter++;
+          } else {
+            // oversize batch
+            noOversizeBatches = false;
+          }
+          elementCounter += currentBatchSize;
+          callback();
+        }
+      })).then(() => {
+        const fullBatches = Math.floor(elements / batchSize);
+        const remainder = elements % batchSize;
+        const partialBatches = remainder > 0 ? 1 : 0;
+        const totalBatches = fullBatches + partialBatches;
+        assert.ok(noOversizeBatches, 'All batches should be less than or equal to batchSize.');
+        assert.strictEqual(batchCounter, totalBatches, 'There should be the correct total number of batches.');
+        assert.strictEqual(fullBatchCounter, fullBatches, 'There should be the correct number of full batches.');
+        assert.strictEqual(partialBatchCounter, partialBatches, 'There should be the correct number of partial batches.');
+        assert.strictEqual(elementCounter, elements, 'There should be the correct number of elements.');
+      });
+    }
+
+    it('single full batch', async function() {
+      return testBatching(5, 5);
+    });
+    it('multiple full batches', async function() {
+      return testBatching(9, 3);
+    });
+    it('only a partial batch', async function() {
+      return testBatching(1, 3);
+    });
+    it('remaining partial batch', async function() {
+      return testBatching(25, 4);
+    });
+  });
+  describe('splitting', async function() {
+    async function testSplitting(elements, batchSize, concurrency) {
+      let elementCounter = 0;
+      return pipeline(Readable.from(Array(elements).keys()), new BatchingStream(batchSize), new SplittingStream(concurrency),
+        new Writable({
+          objectMode: true,
+          write(chunk, encoding, callback) {
+            elementCounter++;
+            callback();
+          }
+        })).then(() => {
+        assert.strictEqual(elementCounter, elements, 'There should be the correct number of elements.');
+      });
+    }
+
+    it('single batch', async function() {
+      return testSplitting(2, 2);
+    });
+
+    it('multiple batches, same size', async function() {
+      return testSplitting(15, 3);
+    });
+
+    it('multiple batches, different size', async function() {
+      return testSplitting(29, 8);
+    });
+
+    it('multiple batches, concurrency', async function() {
```

Comment:
Maybe a test case with multiple batches, different size, and concurrency would be worth it to cover edge cases.